ecs-agent/acs/session/session.go

```diff
@@ -11,13 +11,348 @@
+	taskStopper                    TaskStopper
+	resourceHandler                ResourceHandler
+	backoff                        retry.Backoff
+	sendCredentials                bool
+	clientFactory                  wsclient.ClientFactory
+	metricsFactory                 metrics.EntryFactory
+	minAgentConfig                 *wsclient.WSClientMinAgentConfig
+	addUpdateRequestHandlers       func(wsclient.ClientServer)
+	heartbeatTimeout               time.Duration
+	heartbeatJitter                time.Duration
+	disconnectTimeout              time.Duration
+	disconnectJitter               time.Duration
+	inactiveInstanceReconnectDelay time.Duration
+}
+
+// NewSession creates a new Session.
+func NewSession(containerInstanceARN string,
+	cluster string,
+	discoverEndpointClient api.ECSDiscoverEndpointSDK,
+	credentialsProvider *credentials.Credentials,
+	inactiveInstanceCB func(),
+	clientFactory wsclient.ClientFactory,
+	metricsFactory metrics.EntryFactory,
+	agentVersion string,
+	agentHash string,
+	dockerVersion string,
+	minAgentConfig *wsclient.WSClientMinAgentConfig,
+	payloadMessageHandler PayloadMessageHandler,
+	credentialsManager rolecredentials.Manager,
+	credentialsMetadataSetter CredentialsMetadataSetter,
+	doctor *doctor.Doctor,
+	eniHandler ENIHandler,
+	manifestMessageIDAccessor ManifestMessageIDAccessor,
+	taskComparer TaskComparer,
+	sequenceNumberAccessor SequenceNumberAccessor,
+	taskStopper TaskStopper,
+	resourceHandler ResourceHandler,
+	addUpdateRequestHandlers func(wsclient.ClientServer),
+) Session {
+	backoff := retry.NewExponentialBackoff(connectionBackoffMin, connectionBackoffMax,
+		connectionBackoffJitter, connectionBackoffMultiplier)
+	return &session{
+		containerInstanceARN:           containerInstanceARN,
+		cluster:                        cluster,
+		discoverEndpointClient:         discoverEndpointClient,
+		credentialsProvider:            credentialsProvider,
+		inactiveInstanceCB:             inactiveInstanceCB,
+		clientFactory:                  clientFactory,
+		metricsFactory:                 metricsFactory,
+		agentVersion:                   agentVersion,
+		agentHash:                      agentHash,
+		dockerVersion:                  dockerVersion,
+		minAgentConfig:                 minAgentConfig,
+		payloadMessageHandler:          payloadMessageHandler,
+		credentialsManager:             credentialsManager,
+		credentialsMetadataSetter:      credentialsMetadataSetter,
+		doctor:                         doctor,
+		eniHandler:                     eniHandler,
+		manifestMessageIDAccessor:      manifestMessageIDAccessor,
+		taskComparer:                   taskComparer,
+		sequenceNumberAccessor:         sequenceNumberAccessor,
+		taskStopper:                    taskStopper,
+		resourceHandler:                resourceHandler,
+		addUpdateRequestHandlers:       addUpdateRequestHandlers,
+		backoff:                        backoff,
+		sendCredentials:                true,
```

Comment:
Q: is this value different for fargate?

Reply:
`sendCredentials` is used and behaves the same way across ECS/EC2 and ECS/Fargate launch types (similar to other struct fields such as `backoff`).

Reply:
Cool I guess I am curious if this is ever going to change, but keeping it is fine for me.

Reply:
It gets changed on L255 (i.e., it is `true` for the first connection and `false` for all subsequent connections).